src/components/Books.js

```diff
@@ -0,0 +1,259 @@
+        const outputBooks = [];
+        books.forEach(book => {
+            outputBooks.push({
+                _id: book._id,
+                title: book.title,
+                authorNames: book.authors.map(author => author.name).join(', '),
+                authorIds: book.authors.map(author => author._id),
+                genreNames: book.genres.map(genre => genre.name).join(', '),
+                genreIds: book.genres.map(genre => genre._id),
+                publisherName: book.publisher ? book.publisher.name : null,
+                publisherId: book.publisher ? book.publisher._id : null
+            })
+        })
+        return outputBooks;
+    }
+
+    render() {
+
+        const outputBooks = this.mapBooksToOutputBook(this.props.books);
+       
+        let footer = <div className="p-clearfix" style={{ width: '100%' }}>
+            <Button style={{ float: 'left' }} label="Add" icon="pi pi-plus" onClick={this.addNew} />
+        </div>;
+
+        let dialogFooter = <div className="ui-dialog-buttonpane p-clearfix">
+            {(!this.newBook) && <Button label="Delete" icon="pi pi-times" onClick={this.delete} />}
+            <Button disabled={!this.state.valid} label="Save" icon="pi pi-check" onClick={this.save} />
+        </div>;
+
+        if (this.props.loading) {
+            return <p>loading...</p>
+        }
+
+
+        return (
+
+
+            <div>
+                {this.componentName}
+                <DataTable value={outputBooks} paginator={true} rows={this.props.pageSize} totalRecords={this.props.count}
+                    lazy={true} first={this.state.first} onPage={this.onPage} onSort={this.onSort} loading={this.props.loading} footer={footer}
+                    selectionMode="single" selection={this.state.selectedBook} onSelectionChange={e => this.setState({ selectedBook: e.value })}
+                    onRowSelect={this.onBookSelect}>
+
+                    <Column field="title" header="Title" sortable={true} />
+                    <Column field="authorNames" header="Authors" sortable={true} />
+                    <Column field="genreNames" header="Genres" sortable={true} />
+                    <Column field="publisherName" header="Publisher" sortable={true} />
+                </DataTable>
+
+                <Dialog visible={this.state.displayDialog} width="300px" header="Book Details" modal={true} footer={dialogFooter} onHide={() => this.setState({ displayDialog: false })}>
+                    {
+                        this.state.book &&
+
+                        <div className="p-grid p-fluid">
+                            <div className="p-col-4" style={{ padding: '.75em' }}><label htmlFor="title">Title</label></div>
+                            <div className="p-col-8" style={{ padding: '.5em' }}>
+                                <InputText id="title" required={true} onChange={(e) => { this.updateProperty('title', e.target.value) }} value={this.state.book.title} />
```

Comment:
disable save button when length is < 2. Currently it is sending request and getting error code 400 (BAD REQUEST) back.